xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,203 @@
+      throws IllegalArgumentException {
+    if (this.rbacEngines.size() < 1 || this.rbacEngines.size() > 2) {
+      throw new IllegalArgumentException(
+        "Invalid RBAC list size, must provide either one RBAC or two RBACs. ");
+    } 
+    if (this.rbacEngines.size() == 2 && (this.rbacEngines.get(0).action != Action.DENY 
+        || this.rbacEngines.get(1).action != Action.ALLOW)) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    AuthorizationDecision.Decision authorizationDecision = null;
+    List<String> matchingPolicyNames = new ArrayList<>();
+    List<String> unknownPolicyNames = new ArrayList<>();
+    // Set up activation used in CEL library's eval function.
+    Activation activation = Activation.copyOf(extractFields(args));
+    // Go through each RBAC in the Envoy RBAC list.
+    for (RbacEngine rbacEngine : rbacEngines) {
+      // Go through each condition in the RBAC policy.
+      for (Map.Entry<String, Expr> condition : rbacEngine.conditions.entrySet()) {
+        try {
+          if (matches(condition.getValue(), activation)) {
+            authorizationDecision = rbacEngine.action == Action.ALLOW 
+                ? AuthorizationDecision.Decision.ALLOW : AuthorizationDecision.Decision.DENY;
+            matchingPolicyNames.add(condition.getKey());
+          }
+        } catch (InterpreterException e) {
+          unknownPolicyNames.add(condition.getKey());
+        }
+      }
+      if (matchingPolicyNames.size() > 0) {
+        return new AuthorizationDecision(authorizationDecision, matchingPolicyNames);
+      }
+      if (unknownPolicyNames.size() > 0) {
+        return new AuthorizationDecision(
+            AuthorizationDecision.Decision.UNKNOWN, unknownPolicyNames);
+      }
+    }
+    // No RBAC conditions matched and didn't find unknown conditions.
+    if (this.rbacEngines.size() == 1 && this.rbacEngines.get(0).action == Action.DENY) {
+      return new AuthorizationDecision(
+          AuthorizationDecision.Decision.ALLOW, new ArrayList<String>());
+    }
+    return new AuthorizationDecision(AuthorizationDecision.Decision.DENY, new ArrayList<String>());
+  }
+
+  /** Evaluate if a condition matches the given Enovy Attributes using CEL library. */
+  protected boolean matches(Expr condition, Activation activation) throws InterpreterException {
+    // Set up interpreter used in CEL library's eval function.
+    List<Descriptor> descriptors = new ArrayList<>();
+    RuntimeTypeProvider messageProvider = DescriptorMessageProvider.dynamicMessages(descriptors);
+    Dispatcher dispatcher = DefaultDispatcher.create();
+    Interpreter interpreter = new DefaultInterpreter(messageProvider, dispatcher);
+    // Parse the generated result object to a boolean variable.
+    try {
+      Object result = interpreter.createInterpretable(condition).eval(activation);
```

Comment:
Interpreter will be created for every CEL evaluation checking. Let's move it outside of the loop.